### ISSUE TITLE
Tracking Issue: Remove incremental decoding on ImageDecoder

### ISSUE THREAD

**Issue by fintelia** [image-rs/image#1989]
Nearly all implementations of `ImageDecoder` implement incremental decoding simply by decoding the image all at once and then incrementally copying it out of an in-memory buffer. Since the trait is being refactored to make it object safe for 0.25.x, it is a good time to revisit the incremental decoding interface.

The current plan is to remove the following methods:
```rust
trait ImageDecoder {
    fn scanline_bytes(&self) -> u64;
    fn into_reader(self) -> ImageResult<Self::Reader>;
    fn read_image_with_progress(...) -> ImageResult<()>;
}

trait ImageDecoderRect {
    fn read_rect_with_progress(...) -> ImageResult<()>;
}
```

**If you are relying any of these methods, please comment on this issue and share your use case!**

**Comment by wez** [comment]
Here's how this is being used in wezterm; is there a better way to phrase this?

The purpose of this logic is to be able to decode either apng or png into a series of frames. It's been a while since I wrote this, but I think it looks awkward with the various different color type checks because there wasn't a more convenient method to do the same on the png decoder.

https://github.com/wez/wezterm/blob/05eadb7cbbf26ae073f9555e587a040c82b30928/wezterm-gui/src/glyphcache.rs#L264-L326

**Comment by fintelia** [comment]
@wez you could replace lines 276 and 277 with a call to `decoder.read_image(&mut buf)`. But you can actually do even better by replacing all of lines 272 to 321 with a recently added method:
```rust
let buf = DynamicImage::from_decoder(decoder)?.into_rgba8();
```

**Comment by wez** [comment]
Thanks! I have applied that change and it seems to work

**Comment by CosmicHorrorDev** [comment]
Woops, looks like I saw this a little late :sweat_smile:

I use this feature to stream decoding images into a compressed in-memory blob that gets lazily decompressed if it's accessed (including a transformation that adds in a 0xff alpha channel if one was missing). I'm assuming this is a bit of a niche case since it seems like most people are fine with loading the whole image into memory to do operations, but I should be able to get by with special-casing manual stream decoding using the underlying libraries for a couple of the more common formats

**Comment by maxcountryman** [comment]
It'd be nice to have support for incremental decoding where the underlying format does support it: this enables e.g. streaming responses. So wholesale removal seems a bit unfortunate.

**Comment by fintelia** [comment]
@maxcountryman Could you share more details about your use case?

**Comment by maxcountryman** [comment]
@fintelia I mentioned one in my last comment; at a higher level, avoiding holding all bytes in memory at the same time is a very nice property.

**Comment by fintelia** [comment]
There's a lot of different ways that we could implement streaming/incremental decoding. Do the pixels need to be decoded in scanline order? Or would a low-res copy of the image be more helpful? Does the whole input file have to be available at the start? Are you going to want to decode the whole image or only regions of it? etc.

The old incremental decoding API was added because it "seemed nice to have", but it was actually hard/impossible to implement for many formats, and there wasn't much interest/motivation in working on it without indication that people were actually using it. I'd want to avoid those mistakes if we tried to add back an incremental decoding API some time in the future

**Comment by the8472** [comment]
> at a higher level, avoiding holding all bytes in memory at the same time is a very nice property.

Yes, +1. Firefox has an internal feature called [downscale-during-decode](https://bugzilla.mozilla.org/show_bug.cgi?id=1045926) (it also crops if needed), replicating that when thumbnailing/previewing lots of large images would be useful.

> Do the pixels need to be decoded in scanline order?

This would be helpful, especially when transcoding into a format that supports streaming encoding, then only a tiny slice of pixels would have to be kept in memory.
But in cases where that's not possible any footprint reduction gained from streaming decoding would still be welcome.

> Or would a low-res copy of the image be more helpful?

Yes, that can also be helpful. For example one could decode into a 4x downsampled rect and then perform the final downscale to the target resolution from there, at least in cases where speed is more important than using something fancier than nearest neighbor.

> Does the whole input file have to be available at the start?

Well, streaming input is also desirable.


Basically, those are a pile of related memory-footprint-reduction features¹. If a format or even a particular file supports one or more of them that's great.
If they don't then one can still put that image on a slow path behind a memory semaphore or write a streaming reader out to a `File` before working on it.

¹ those potentially orthogonal features being

- streaming input
- scanline order streaming decode
- downsampling / downscaling during decode
- crop during decode

**Comment by fintelia** [comment]
The challenge is going from "nice to have" to a concrete use-case that justifies the implementation and ongoing maintenance effort _likely at the expense of other features_.

**Comment by the8472** [comment]
Well, it's essentially about performance and resource consumption.

If you need to decode a whole high-resolution image for processing that gobbles up RAM which means you need to add resource limits, to process fewer images at a time. On resource-constrained environments it may not be possible at all to handle things above certain resolutions.

This comes up when writing image galleries, thumbnailers, doing bulk processing or just handling very large images. Pretty much the reasons why firefox added DDD.

**Comment by ender672** [comment]
I have a use case to share. I work with image processing tools, and I have large PNG and JPEG files on my computer. One in particular is `spark.png` from https://www.bamsoftware.com/hacks/deflate.html .

The simple act of copying `spark.png` into a directory and opening nautilus causes my system to freeze until eventually the OOM killer starts killing off processes (if I'm lucky it leaves my gnome-shell process running and I don't lose all my windows):
```
$ cp ../spark.png .
$ nautilus .
** Message: 11:01:09.868: Connecting to org.freedesktop.Tracker3.Miner.Files
MESA-INTEL: warning: ../src/intel/vulkan/anv_formats.c:993: FINISHME: support more multi-planar formats with DRM modifiers
MESA-INTEL: warning: ../src/intel/vulkan/anv_formats.c:959: FINISHME: support YUV colorspace with DRM format modifiers
Terminated                 nautilus .
```
It would be nice if image formats that support scanline-at-a-time reading (non-interlaced PNG, jpeg, etc) could be decoded and downscaled one scanline at a time. It saves memory and, when done right, is incredibly performant.

The `png` crate has https://docs.rs/png/0.18.1/png/struct.Reader.html#method.next_row . For jpeg images, I'm using libjpeg-turbo with FFI to decode one scanline at a time.

I created a rust crate for this purpose and with the hopes of fixing nautilus, but I'm stuck on this image-rs limitation.

**Comment by 197g** [comment]
*Thank you*. That is something to work on, it thankfully eliminates some of the API designs outright. Unfortunately maybe too many. We probably don't want `scanline` to exist on all formats again, that fiction simply does not work for some ways of compression. Tiled formats could be made to buffer a bands at a time but that can't be the _basis_ of their generic reader implementation; it'd be a separate way of decoding them where you eat some double-buffering costs. It's very valuable to say that memory-pressure is however another major constraint here since that questions if even that is acceptable. (Well apparently its workable for you with libjpeg that is such a format). Although you may prefer seeing that band of scanlines at once? But then you look at layers (zero-duration frames) in HEIF and the anti-ringing filters to be applied across blocks and even the fiction of a simple tiling solution goes out the window.

The real problem for me is still abstraction. It works with individual formats; that's great. The list where it does is hardly my concern—the unknowns where it does _not_ is. Opening issues in the upstream crates for formats that do not yet have a mode but could is a form of progress—we might have a tracking issue for it to also have an overview of approaches. But even with those getting addressed there would remain a _massive_ task of a uniform interface to all those different variants. Whether the code for bindings lives here or in the application is no difference but it's a big ask to maintain something overly generic. We would at least need a way to _incrementally_ get to a point of support (some experimental API that we are comfortable enough ripping out if it does not scale).

To get one concrete question out of this: How would you handle the case where a decoder simply does/can not provide any option for memory-bounded partial decoding? We very likely won't make inclusion dependent on it; definitely not for those in `image-extra`. So it'd be interesting how critical the interface must treat the lack of this capability.

**Comment by ender672** [comment]
After reading through how Natilus and `glycin` uses `image`, I'm no longer convinced that `image` needs a public `scanline` API. It seems that if a consumer wants that level of control, they might as well directly use a decoder library. I no longer think it's the role of `image` to provide this API since it has its own resizer.

Instead, what I think `glycin` and most consumers will want is a way to indicate what image scaling should happen before the image is fully decoded. When memory matters, image decoding and scaling are linked. Unfortunately, as you pointed out, how to memory-efficiently do decode+scale looks very different depending on the compression, and sometimes must resort to full image. I'm guessing the idea of decode+scale will give you a headache since it's nicer to treat them separately.

As for how to shape such an API, I really don't know. Since I gave a use case of glycin with a huge PNG, I would start with that. You have a PNG where the decoder *does* support scanline-at-a-time decoding, and a caller that knows what max-dimensions it wants for thumbnailing. Seems like the caller should be allowed to tell `image` "I want this decoded into a 100x100 box" and they get a pixel buffer just like before, just using a lot less memory. The naive and specific solution is usually the best start. Do this for every format that supports partial decode, and maybe a uniform interface will become evident.

As to your last question, if an image format does not support partial decoding, it seems Nautilus/glycin will need to decide what to do. Bite the bullet and put the full image in memory, or cancel and show the user a placeholder. I suppose it is the role `image` to give a hint on what it can do, e.g. "decoding will use XXMB memory" or "this image can do scanline-at-a-time".

**Comment by fintelia** [comment]
@ender672 I'm surprised to hear that nautilus is triggering OOM on large images. We already expose APIs to see how much memory is required to store an output image (or depending on the API to trigger an error if a user selected limit is exceeded). I'd expect it to refuse to generate thumbnails for images that need more than, say, 1GB or so. 

That said, I do like the idea of having an API to decode and downsample simultaneously. That's more compatible with things like interlaced PNG (which gives you a 1/8th x 1/8th resolution to start with and then provides more details in subsequent passes) and JPEG (which encodes pixels in 8x8 blocks). It would require implementation for each decoder, but I think there's a bunch that could give downsampled output but would struggle to do scanline-at-a-time.
